// meta-bolt/dist/lib.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pause = exports.posix = exports.safeCreate = exports.removeIfExists = exports.sep = exports.resetLog = exports.log = exports.conColors = void 0;
const os = __importStar(require("os"));
const fs = __importStar(require("fs"));
exports.conColors = {
    white: "\u001b[0m",
    whiteBold: "\u001b[1;37m",
    cyan: "\u001b[1;36m",
    purple: "\u001b[1;35m",
    blue: "\u001b[1;34m",
    yellow: "\u001b[1;33m",
    green: "\u001b[1;32m",
    red: "\u001b[1;31m",
    grey: "\u001b[1;30m",
};
const log = (str, succeeded, info) => {
    const res = succeeded ? "succeeded √" : "failed ×";
    const color = succeeded ? exports.conColors.cyan : exports.conColors.red;
    console.log(`${color}${str} → ${res} ${(info && ":") || ""} ${info || ""}`);
    if (!succeeded)
        throw info;
};
exports.log = log;
const resetLog = () => {
    console.log(exports.conColors.white);
};
exports.resetLog = resetLog;
exports.sep = os.platform() == "win32" ? "\\" : "/";
const removeIfExists = (dir) => {
    try {
        fs.existsSync(dir) && fs.unlinkSync(dir);
        return [true, "Removed"];
    }
    catch (e) {
        return [false, e];
    }
};
exports.removeIfExists = removeIfExists;
const safeCreate = (dir) => {
    try {
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir);
            return [true, "Created"];
        }
        return [true, "Already Exists"];
    }
    catch (e) {
        return [false, e];
    }
};
exports.safeCreate = safeCreate;
const posix = (str) => str.replace(/\\/g, "/");
exports.posix = posix;
const pause = (ms) => {
    return new Promise((resolve) => {
        setTimeout(() => resolve(""), ms);
    });
};
exports.pause = pause;
